Add 'show' and 'clear' command for PG drop (internal review)

ayurkiv-nvda opened the pull request:
<!--
    Please make sure you've read and understood our contributing guidelines:
    https://github.com/Azure/SONiC/blob/gh-pages/CONTRIBUTING.md

    ** Make sure all your commits include a signature generated with `git commit -s` **

    If this is a bug fix, make sure your description includes "closes #xxxx",
    "fixes #xxxx" or "resolves #xxxx" so that GitHub automatically closes the related
    issue when the PR is merged.

    If you are adding/modifying/removing any command or utility script, please also
    make sure to add/modify/remove any unit tests from the tests
    directory as appropriate.

    If you are modifying or removing an existing 'show', 'config' or 'sonic-clear'
    subcommand, or you are adding a new subcommand, please make sure you also
    update the Command Line Reference Guide (doc/Command-Reference.md) to reflect
    your changes.

    Please provide the following information:
-->

#### What I did
I have added possibility to view and clear PG dropped packet stat

#### How I did it
I have added script which can generate table that shows current PG dropped packet statistic and also can clean it
Clean and Show is implemented with using serialized stats from previous execution.
To calculate current stats we do next step:
1) take old serialized data (if no 'clear' executed all serialized data is 0)
2) take current stat
3) show diff between current and old

#### How to verify it
run
**show priority-group drop counters** - show current stat
**sonic-clear priority-group drop counters** - clear current stat

#### Previous command output (if the output of a command-line utility has changed)

#### New command output (if the output of a command-line utility has changed)
**show priority-group drop counters** - show current stat
it uses **pg-drop -c show** - new script with command
**sonic-clear priority-group drop counters** - clear current stat
it uses **pg-drop -c clear**

ayurkiv-nvda committed:
Add 'show' and 'clear' command for PG drop (internal review)

liat-grozovik commented on clear/main.py:
```diff
@@ -216,6 +216,17 @@ def clear_wm_pg_shared():
+
+@drop.command('counters')
+def clear_pg_counters():
+    """Clear PG dropped packets stats counter """
```
why do we use sometimes PG and some time priority group? lets use one terminology in both UI and code
may i suggest to remove the stats? it has nothing to do with stats but these are counters.

liat-grozovik commented on scripts/pg-drop:
```diff
@@ -0,0 +1,233 @@
+#!/usr/bin/env python3
```
this is also needed for 201911. as this is python3 does it mean you will have to create a dedicated PR for 201911 which is not python 3?

ayurkiv-nvda replied:
yes, I will create dedicated PR for 201911

liat-grozovik commented on scripts/pg-drop:
```diff
@@ -0,0 +1,233 @@
+    dropstat_dir_prefix = '/tmp/dropstat'
+    return "{}-{}/".format(dropstat_dir_prefix, os.getuid())
+
+class PGdropstat(object):
```
is there a convention where each letter of a word in the name should be in capital letter such as PgDropStat?
please check if all other counters use Stat and not Counter/Count

liat-grozovik commented on scripts/pg-drop:
```diff
@@ -0,0 +1,233 @@
+            self.port_pg_map[port] = {}
+            self.port_name_map[self.counter_port_name_map[port]] = port
+
+
```
remove extra empty line

liat-grozovik commented on scripts/pg-drop:
```diff
@@ -0,0 +1,233 @@
+                               "header_prefix": "PG"},
+        }
+
+    def get_pg_index(self, table_id):
```
i am not sure i understand the concept of table. do we have more than 1 table for the PG? what are the expected tables?

ayurkiv-nvda replied:
**table_id** - is actually object id  for entry in redis db. 
For example **oid:0x1a000000000121** - is our id
**COUNTERS:oid:0x1a000000000121** - full path for id
**redis-cli -n 2  HGET  COUNTERS:oid:0x1a000000000121 SAI_INGRESS_PRIORITY_GROUP_STAT_DROPPED_PACKETS**- command for retrieving value

liat-grozovik commented on scripts/pg-drop:
```diff
@@ -0,0 +1,233 @@
+
+    def get_counters(self, table_prefix, port_obj, idx_func, counter_name):
+        """
+            Get the counters from specific table.
```
'Get counters of a specific table'

liat-grozovik commented on scripts/pg-drop:
```diff
@@ -0,0 +1,233 @@
+        if os.path.isfile(self.port_drop_stats_file):
+            port_drop_ckpt = pickle.load(open(self.port_drop_stats_file, 'rb'))
+
+        # header list contains the port name followed by the pgs. fields is used to populate the pg values
```
PGs instead of pgs. PG instead of pg
Use capital letters when needed

liat-grozovik commented on scripts/pg-drop:
```diff
@@ -0,0 +1,233 @@
+        print(type["message"])
+        print(tabulate(table, self.header_list, tablefmt='simple', stralign='right'))
+
+
```
remove empty line

liat-grozovik commented on scripts/pg-drop:
```diff
@@ -0,0 +1,233 @@
+            current_stat_dict[obj] = self.get_counts(counters, counter_object_name_map[obj])
+        return current_stat_dict
+
+
```
remove empty line

liat-grozovik commented on scripts/pg-drop:
```diff
@@ -0,0 +1,233 @@
+                                     formatter_class=argparse.RawTextHelpFormatter,
+                                     epilog="""
+Examoles:
+pg_drop -c show
```
do we usually use _ or - ? lets make sure we are aligned as after that it is difficult to change.

liat-grozovik commented on scripts/pg-drop:
```diff
@@ -0,0 +1,233 @@
+pg_drop -c clear
+""")
+
+
```
empty line?

liat-grozovik commented on show/main.py:
```diff
@@ -605,6 +605,17 @@ def wm_pg_shared():
+
+@drop.command('counters')
+def pg_drop_counters():
+    """Show dropped packets for pg"""
```
priority-group instead of pg also in comment

liat-grozovik reviewed (changes requested):
- Please confirm show command does not need sudo. if so, need to include it in the show command execution.
- Please add comments to functions of description, input, output, etc.
- missing command line reference update of the new commands

ayurkiv-nvda commented on clear/main.py:
```diff
@@ -216,6 +216,17 @@ def clear_wm_pg_shared():
 
+@priority_group.group()
+def drop():
+    """Clear priority group dropped packets stats"""
```
PG

liat-grozovik replied:
priority-group

ayurkiv-nvda commented on scripts/pg-drop:
```diff
@@ -0,0 +1,233 @@
+
+#####################################################################
+#
+# pg_drop is a tool for displaying ingress pg dropped packet stats.
```
rename show/clear

volodymyrsamotiy commented on scripts/pg-drop:
```diff
@@ -0,0 +1,233 @@
+class PGdropstat(object):
+
+    def __init__(self):
+        self.counters_db = swsssdk.SonicV2Connector(host='127.0.0.1')
```
Use "swsscommon" instead of "swsssdk" since it is deprecated.
Check recent changes in "watermarkstat" script as a reference.

ayurkiv-nvda commented on scripts/pg-drop:
```diff
@@ -0,0 +1,233 @@
+            self.port_pg_map[port][pg] = counter_pg_name_map[pg]
+
+        self.pg_drop_types = {
+            "pg_drop"       : {"message" : "Ingress dropped packets PG:",
```
rename ingress pg dropped packets

volodymyrsamotiy commented on scripts/pg-drop:
```diff
@@ -0,0 +1,233 @@
+
+    def get_counts(self, counters, oid):
+            """
+                Get the pg drop counts for an individual counter.
```
pg -> PG

volodymyrsamotiy commented on scripts/pg-drop:
```diff
@@ -0,0 +1,233 @@
+
+    def get_counts_table(self, counters, object_table):
+        """
+            Returns a dictionary containing a mapping from an object (like a port)
```
pg -> PG

ayurkiv-nvda commented on scripts/pg-drop:
```diff
@@ -0,0 +1,233 @@
+
+    def clear_drop_counts(self):
+        """
+            Clears the current pg drop counter.
```
pg -> PG

ayurkiv-nvda commented on scripts/pg-drop:
```diff
@@ -0,0 +1,233 @@
+    parser = argparse.ArgumentParser(description='Display pg drop counter',
+                                     formatter_class=argparse.RawTextHelpFormatter,
+                                     epilog="""
+Examoles:
```
examples

ayurkiv-nvda committed:
Add description for new CLI (pg-drop)

liat-grozovik commented:
@ayurkiv-nvda i see you changed all places to PG but for watermark we use priority-group

def pg_drop_counters():
    """Show dropped packets for PG"""

@priority_group.group(name='persistent-watermark')
def persistent_watermark():
    """Show priority-group persistent WM"""

why the inconsistency?

for the user we should use priority-group inside the code we can use pg

ayurkiv-nvda commented:
> * Please confirm show command does not need sudo. if so, need to include it in the show command execution.
confirmed, no sudo
> * Please add comments to functions of description, input, output, etc.
added in new commit
> * missing command line reference update of the new commands
added in new commit in Command-Reference.md file

ayurkiv-nvda commented:
> @ayurkiv-nvda i see you changed all places to PG but for watermark we use priority-group
> 
> def pg_drop_counters():
> """Show dropped packets for PG"""
> 
> @priority_group.group(name='persistent-watermark')
> def persistent_watermark():
> """Show priority-group persistent WM"""
> 
> why the inconsistency?
> 
> for the user we should use priority-group inside the code we can use pg

Done

Commit:
Add UT for show/clear
